runtime/src/main/java/io/quarkiverse/messaginghub/pooled/jms/PooledJmsDecorator.java

```diff
@@ -32,46 +35,66 @@ private ConnectionFactory getConnectionFactory() {
+        return factory;
     }
 
     @Override
     public Connection createConnection() throws JMSException {
-        return getConnectionFactory().createConnection();
+        return isInJmsPoolConnectionFactory() ? delegate.createConnection() : getConnectionFactory().createConnection();
     }
 
     @Override
     public Connection createConnection(final String userName, final String password) throws JMSException {
-        return getConnectionFactory().createConnection(userName, password);
+        return isInJmsPoolConnectionFactory() ? delegate.createConnection(userName, password)
+                : getConnectionFactory().createConnection(userName, password);
     }
 
     @Override
     public JMSContext createContext() {
-        return runInContext(ConnectionFactory::createContext);
+        return getConnectionFactory().createContext();
     }
 
     @Override
     public JMSContext createContext(final String userName, final String password) {
-        return runInContext(cf -> cf.createContext(userName, password));
+        return getConnectionFactory().createContext(userName, password);
     }
 
     @Override
     public JMSContext createContext(final String userName, final String password, final int sessionMode) {
-        return runInContext(cf -> cf.createContext(userName, password, sessionMode));
+        return getConnectionFactory().createContext(userName, password, sessionMode);
     }
 
     @Override
     public JMSContext createContext(final int sessionMode) {
-        return runInContext(cf -> cf.createContext(sessionMode));
+        return getConnectionFactory().createContext(sessionMode);
     }
 
-    private JMSContext runInContext(Function<ConnectionFactory, JMSContext> process) {
-        ConnectionFactory cf = getConnectionFactory();
-        isContext = true;
-        try {
-            return process.apply(cf);
-        } finally {
-            isContext = false;
-        }
+    @Override
+    public XAConnection createXAConnection() throws JMSException {
+        return isInJmsPoolConnectionFactory() ? delegate.createXAConnection()
+                : ((XAConnectionFactory) getConnectionFactory()).createXAConnection();
+    }
+
+    @Override
+    public XAConnection createXAConnection(final String userName, final String password) throws JMSException {
+        return isInJmsPoolConnectionFactory() ? delegate.createXAConnection(userName, password)
+                : ((XAConnectionFactory) getConnectionFactory()).createXAConnection(userName, password);
+    }
+
+    @Override
+    public XAJMSContext createXAContext() {
+        return ((XAConnectionFactory) factory).createXAContext();
+    }
+
+    @Override
+    public XAJMSContext createXAContext(final String userName, final String password) {
+        return ((XAConnectionFactory) factory).createXAContext(userName, password);
+    }
+
+    private boolean isInJmsPoolConnectionFactory() {
+        return StackWalker.getInstance().walk(
```

Comment:
I think that you can share a `StackWalker` instance to conserve the memory, it's thread-safe...

Reply:
Hmm, but `StackWalker.getInstance()` seems returning a static `DEFAULT_WALKER` which I think it should be a singleton one.

Reply:
You're right, feel free to ignore my stupid comment ;-)